src/common/device_helpers.cuh

```diff
@@ -391,13 +408,19 @@ struct XGBDefaultDeviceAllocatorImpl : thrust::device_malloc_allocator<T> {
     GlobalMemoryLogger().RegisterDeallocation(ptr.get(), n * sizeof(T));
     return SuperT::deallocate(ptr, n);
   }
+#if defined(XGBOOST_USE_RMM) && XGBOOST_USE_RMM == 1
+  using cuda_mr = rmm::mr::cuda_memory_resource;
+  using pool_mr = rmm::mr::pool_memory_resource<cuda_mr>;
+  XGBDefaultDeviceAllocatorImpl() : SuperT(new pool_mr(new cuda_mr), cudaStream_t{0}) {}
+#endif  // defined(XGBOOST_USE_RMM) && XGBOOST_USE_RMM == 1
 };
 
 /**
- * \brief Caching memory allocator, uses cub::CachingDeviceAllocator as a back-end and logs allocations if verbose. Does not initialise memory on construction.
+ * \brief Caching memory allocator, uses cub::CachingDeviceAllocator as a back-end and logs
+ * allocations if verbose. Does not initialise memory on construction.
  */
 template <class T>
-struct XGBCachingDeviceAllocatorImpl : thrust::device_malloc_allocator<T> {
+struct XGBCachingDeviceAllocatorImpl : XGBBaseDeviceAllocator<T> {
```

Comment:
What allocations go through `XGBDefaultDeviceAllocatorImpl` vs. `XGBCachingDeviceAllocatorImpl`?

I suspect the reason your tests are going OOM are because these two things are conflicting with each other. This caching device allocator isn't using the pool resource, so it's allocations are going to conflict with the allocation of the pool.

Reply:
`XGBCachingDeviceAllocatorImpl` is used for small, temporary allocations needed for various algorithm kernels. Let me try replacing the CUB allocator.

Reply:
Another thing you can try is increasing the size of the initial pool. By default it uses ~1/2 of available GPU memory. Depending on your allocation pattern, you could end up fragmented such that even if `n` bytes are "free" in the pool, there may not be a contiguous chunk of `n` bytes to satisfy your allocation. Using a large initial pool helps reduce fragmentation issues.

Reply:
Replacing `XGBCachingDeviceAllocatorImpl` with the RMM pool allocator didn't seem to help. Maybe I'm using `rmm::mr::thrust_allocator` incorrectly. Can you point me to a working example with the Thrust allocator?

Reply:
I replaced `new pool_mr(new cuda_mr)` with `rmm::mr::get_default_resource()` (per advice of @harrism) and now the crash is gone. I reverted `XGBCachingDeviceAllocatorImpl` back to the CUB allocator because some CUB algorithms crashed when they were given a device vector allocated with RMM pool allocator.